pkg/codec/by_item_type_modifier.go

```diff
@@ -0,0 +1,52 @@
+package codec
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+)
+
+// NewByItemTypeModifier returns a Modifier that uses modByItemType to determine which Modifier to use for a given itemType.
+func NewByItemTypeModifier(modByItemType map[string]Modifier) (Modifier, error) {
+	if modByItemType == nil {
+		modByItemType = map[string]Modifier{}
+	}
+
+	return &byItemTypeModifier{
+		modByitemType: modByItemType,
+	}, nil
+}
+
+type byItemTypeModifier struct {
+	modByitemType map[string]Modifier
+}
+
+func (b *byItemTypeModifier) RetypeForOffChain(onChainType reflect.Type, itemType string) (reflect.Type, error) {
+	mod, ok := b.modByitemType[itemType]
+	if !ok {
+		return nil, fmt.Errorf("%w: cannot find modifier for %s", types.ErrInvalidType, itemType)
+	}
+
+	return mod.RetypeForOffChain(onChainType, itemType)
+}
+
+func (b *byItemTypeModifier) TransformForOnChain(offChainValue any, itemType string) (any, error) {
```

Comment:
Would these methods read a little more clearly if they used `To` instead of `For`?
```suggestion
func (b *byItemTypeModifier) TransformToOnChain(offChainValue any, itemType string) (any, error) {
```

Reply:
Sure, I'll rename Retype with To also then :).